Test/lambdas/MatrixAssoc.dfy

```diff
@@ -0,0 +1,270 @@
+// RUN: %dafny /compile:0 /dprint:"%t.dprint" "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+type Pos = x | 0 < x witness 1
+
+/** Matrix dimension */
+const N : Pos
+
+/** Matrix indices */
+type Index = x | 0 <= x < N      // matrix indices
+
+type Matrix = Index -> Index -> int
+
+/** Function extensionality */
+lemma FunEq<X, Y>(f: X -> Y, g: X -> Y)
+    requires forall x :: f(x) == g(x)
+    ensures f == g
+{ assume false; }
+
+lemma Same(a: Matrix, b: Matrix)
+    requires forall x, y :: a(x)(y) == b(x)(y)
+    ensures a == b
+{
+    forall x ensures a(x) == b(x) {
+        FunEq(a(x), b(x));
+    }
+    FunEq(a, b);
+}
+
+/** Σ from i to n of f(i) */
```

Comment:
The comment should say "`0` to `n`" (which leaves `i` rather unbound in the sentence). Perhaps

    Σ f(i) where i ranges from 0 to n

On the subject of "`0` to `n`":  One way to distinguish half-open intervals (like `[0..100]` in Dafny, `[0, 100)` in textbooks, or `i: 0 <= i < 100` using a predicate) from closed intervals (like `[0, 99]` in textbooks or `i: 0 <= i <= 99`) is to use "`0` to `100`" for the former and "`0` through `99`" for the latter. The mnemonic for these is that "to" tells you to stop as soon as you reach `100`, whereas "through" says you go all the way through `99` before you stop. Using this convention (which I'm not claiming is widespread, but it makes sense to me), "`0` to `n`" is indeed what you want. But if you want to emphasize that more, you could say (the grammatically dubious, but unambiguous) "`0` to less than `n`".